lib/pages/search/server_search_controller.dart

```diff
@@ -14,34 +17,71 @@ class ServerSearchController with SearchDebouncerMixin {
     Right(ServerSearchInitial()),
   );
 
-  String keyword = '';
+  final isLoadingMoreNotifier = ValueNotifier<bool>(false);
+
+  final scrollController = ScrollController();
+
+  StreamSubscription? _searchSubscription;
 
   void init() {
     initializeDebouncer((searchTerm) {
       if (searchTerm.isNotEmpty) {
-        keyword = searchTerm;
         searchUnencryptedMessages(searchTerm);
       } else {
         serverSearchNotifier.value = Right(ServerSearchInitial());
       }
     });
+    scrollController.addLoadMoreListener(loadMore);
   }
 
   void dispose() {
     super.disposeDebouncer();
+    scrollController.dispose();
+    serverSearchNotifier.dispose();
+    _searchSubscription?.cancel();
+    isLoadingMoreNotifier.dispose();
   }
 
   void searchUnencryptedMessages(String searchTerm) {
-    _serverSearchInteractor
+    _searchSubscription = _serverSearchInteractor
         .execute(
       searchTerm: searchTerm,
     )
-        .listen((searchResult) {
-      serverSearchNotifier.value = searchResult;
-    });
+        .listen(
+      (searchResult) {
+        serverSearchNotifier.value = searchResult;
+      },
+      onDone: () {
+        scrollController.tryLoadMoreIfNeeded(loadMore);
+      },
+    );
   }
 
   void onSearchBarChanged(String keyword) {
     setDebouncerValue(keyword);
   }
+
+  void loadMore() {
+    final lastSuccess =
+        serverSearchNotifier.value.getSuccessOrNull<ServerSearchChatSuccess>();
```

Comment:
IMO, it is not a good idea. I worry about other state will stay in this notification. What do you think if you keep the `search criteria` in separate with the state?

Reply:
why not, you can load more only when current state is success, then I use data from last success to query more

Reply:
Because, we don't know exactly in the future how this `notifier` will be updated. Other interactor, parallel use-case, ..... these maybe make this function can not work.

Reply:
new state should also keep keyword and next batch, it required for load more data
next batch is the thing need to check load more available, not last success

Reply:
I agreed with @hoangdat

Reply:
keep in mind, only trust source, keep data to fetch more in another variable will make hard to maintain, and make side effect

Reply:
if that, `serverSearchNotifier` only serve for the case of server side search?

Reply:
if you want add local search, we have two case
- there are separated list => answer is YES, easy to understand
- only one list for both local and server search => answer is NO, serverSearchNotifier will keep state pushed by interactor, not only server search, but also local search, and interactor also use last state to handle load more.

Reply:
please discuss it tmr:
- how to add other filter (Ex: date)?